### PR TITLE
fix(cedarling): Make cedarling trusted store compatible with agama lab

### LINKED ISSUE #10038: fix(cedarling): Make cedarling Policy Store compatible with Agama Lab
We need to make trusted store compatible with agama-lab. See [Wiki](https://github.com/JanssenProject/jans/wiki/Cedarling-Nativity-Plan#cedarling-policy-store) for final design.

### PR DESCRIPTION
**Pull request by olehbozhok**
### Prepare

- [x] Read [PR guidelines](https://github.com/JanssenProject/jans/blob/main/docs/CONTRIBUTING.md#prs)
- [x] Read [license information](https://github.com/JanssenProject/jans/blob/main/LICENSE)

-------------------

### Description
fix loading cedarling trusted issuer from agama lab

#### Target issue
  
  
  
  
closes #10038

#### Implementation Details
  

-------------------
### Test and Document the changes
- [x] Static code analysis has been run locally and issues have been fixed
- [x] Relevant unit and integration tests have been added/updated
- [ ] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)

Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [x] **I confirm that there is no impact on the docs due to the code changes in this PR.**

### CODE REVIEW

**Review by rmarinn — DISMISSED** [review]

File: jans-cedarling/cedarling/src/init/policy_store.rs [hunk 1]
```diff
@@ -42,7 +44,27 @@ pub(crate) fn load_policy_store(
...
+    let policy_store = match serde_json::from_str::<PolicyStore>(policies_json) {
+        Ok(policy_store) => policy_store,
+        Err(err) => {
+            // try to decode compatible to agama-lab
+            let Ok(result_map) =
+                serde_json::from_str::<HashMap<String, PolicyStore>>(policies_json)
+            else {
+                // return previous error to be unit test compatible
+                return Err(err);
+            };
```
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/init/policy_store.rs]: in Rust, it's better to use `match` then `if else`
```
let result_map = match serde_json::from_str::<HashMap<String, PolicyStore>>(policies_json) {
    Ok(map) => map,
    Err(_) => return Err(err), // Return the original error here
};
```

File: jans-cedarling/cedarling/src/init/policy_store.rs [hunk 2]
```diff
@@ -42,7 +44,27 @@ pub(crate) fn load_policy_store(
...
+            if result_map.len() != 1 {
+                return Err(serde::de::Error::custom(
+                    "currently we support only one policy store",
+                ));
+            }
+            result_map
+                .into_values()
+                .next()
+                .expect("value should be present in the iterator")
```
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/init/policy_store.rs]: Using `.expect("value should be present in the iterator")`  can lead to a panic if, for some reason,  result_map.into_values().next() doesn't return a value. Although, unlikely given the checks above, it's generally good practice to handle all possible error cases more gracefully. For instance, replacing `expect` with `ok_or_else` could make this safer: 
```
result_map.into_values().next().ok_or_else(|| serde::de::Error::custom("expected one policy store entry"))
```

File: jans-cedarling/cedarling/src/init/service_config.rs [hunk 3]
```diff
@@ -50,6 +50,19 @@ impl ServiceConfig {
...
+            .identity_source
+            .clone()  // we need clone to avoid borrowing
+            .unwrap_or_default()
+            .values()
+            .map(|identity_source| {
+                // TODO This will need to be generalised.
+                // TrustedIssuerAndOpenIdConfig::fetch(identity_source.clone(), &client)
+                Ok(identity_source.openid_configuration_endpoint.clone())
+            })
+            .collect::<Result<Vec<String>, jwt::KeyServiceError>>()?;
```
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/init/service_config.rs:64]: If you don't want to stop the iteration because of any error and propagate up immediately, consider using `partition` to categorize them, with something like: 
```
        let (successful_results, errors): (Vec<_>, Vec<_>) = policy_store
        .identity_source
        .clone()  // Clone if necessary, as in your original code
        .unwrap_or_default()
        .values()
        .map(|identity_source| {
            // Replace this with the actual fetch if needed, for now just clone the endpoint
            Ok::<String, jwt::KeyServiceError>(identity_source.openid_configuration_endpoint.clone())
        })
        .partition(Result::is_ok);
```

File: jans-cedarling/cedarling/src/tests/agama_policy_test.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
...VudGl0eU9yQ29tbW9uIiwibmFtZSI6IkxvbmcifSwiaWF0Ijp7InR5cGUiOiJFbnRpdHlPckNvbW1vbiIsIm5hbWUiOiJMb25nIn0sImlzcyI6eyJ0eXBlIjoiRW50aXR5T3JDb21tb24iLCJuYW1lIjoiVHJ1c3RlZElzc3VlciJ9LCJqdGkiOnsidHlwZSI6IkVudGl0eU9yQ29tbW9uIiwibmFtZSI6IlN0cmluZyJ9fX19LCJpZF90b2tlbiI6eyJzaGFwZSI6eyJ0eXBlIjoiUmVjb3JkIiwiYXR0cmlidXRlcyI6eyJhY3IiOnsidHlwZSI6IkVudGl0eU9yQ29tbW9uIiwibmFtZSI6IlN0cmluZyJ9LCJhbXIiOnsidHlwZSI6IkVudGl0eU9yQ29tbW9uIiwibmFtZSI6IlN0cmluZyJ9LCJhdWQiOnsidHlwZSI6IkVudGl0eU9yQ29tbW9uIiwibmFtZSI6IlN0cmluZyJ9LCJleHAiOnsidHlwZSI6IkVudGl0eU9yQ29tbW9uIiwibmFtZSI6IkxvbmcifSwiaWF0Ijp7InR5cGUiOiJFbnRpdHlPckNvbW1vbiIsIm5hbWUiOiJMb25nIn0sImlzcyI6eyJ0eXBlIjoiRW50aXR5T3JDb21tb24iLCJuYW1lIjoiVHJ1c3RlZElzc3VlciJ9LCJqdGkiOnsidHlwZSI6IkVudGl0eU9yQ29tbW9uIiwibmFtZSI6IlN0cmluZyJ9LCJzdWIiOnsidHlwZSI6IkVudGl0eU9yQ29tbW9uIiwibmFtZSI6IlN0cmluZyJ9fX19LCJJc3N1ZSI6eyJzaGFwZSI6eyJ0eXBlIjoiUmVjb3JkIiwiYXR0cmlidXRlcyI6eyJjb3VudHJ5Ijp7InR5cGUiOiJFbnRpdHlPckNvbW1vbiIsIm5hbWUiOiJTdHJpbmcifSwib3JnX2lkIjp7InR5cGUiOiJFbnRpdHlPckNvbW1vbiIsIm5hbWUiOiJTdHJpbmcifX19fSwiUm9sZSI6e319LCJhY3Rpb25zIjp7IlVwZGF0ZSI6eyJhcHBsaWVzVG8iOnsicmVzb3VyY2VUeXBlcyI6WyJJc3N1ZSJdLCJwcmluY2lwYWxUeXBlcyI6WyJXb3JrbG9hZCIsIlVzZXIiLCJSb2xlIl19fX19fQ==",
```
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/tests/agama_policy_test.rs]: It could be more human readable to put the decoded version as comment

File: jans-cedarling/cedarling/src/tests/agama_policy_test.rs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+#[test]
+fn load_agama_policy_store_json() {
+    let agama_policy_store = serde_json::json!({
+        "fbcd361d3243f4808d5e6f89ca81eaf4ca03900b407e": {
+            "cedar_version": "v2.4.7",
+            "policies": {
+                "840da5d85403f35ea76519ed1a18a33989f855bf1cf8": {
+                    "description": "simple policy example for principal workload",
+                    "creation_date": "2024-09-20T17:22:39.996050",
+                    "policy_content": "cGVybWl0KAogICAgcHJpbmNpcGFsIGlzIEphbnM6Oldvcmtsb2FkLAogICAgYWN0aW9uIGluIFtKYW5zOjpBY3Rpb246OiJVcGRhdGUiXSwKICAgIHJlc291cmNlIGlzIEphbnM6Oklzc3VlCil3aGVuewogICAgcHJpbmNpcGFsLm9yZ19pZCA9PSByZXNvdXJjZS5vcmdfaWQKfTs="
```
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/tests/agama_policy_test.rs]: It could be more human readable to put the decoded version as comment

File: jans-cedarling/cedarling/src/init/policy_store.rs (hunk 1 above)
Comment (olehbozhok) [reply jans-cedarling/cedarling/src/init/policy_store.rs]: when we ignore error it is OK to use 
```
let OK(_) = some else{
...
}
```
https://rust-lang.github.io/rfcs/3137-let-else.html#guide-level-explanation

Comment (olehbozhok) [reply jans-cedarling/cedarling/src/init/policy_store.rs]: and also https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html

File: jans-cedarling/cedarling/src/init/policy_store.rs (hunk 2 above)
Comment (olehbozhok) [reply jans-cedarling/cedarling/src/init/policy_store.rs]: this code will be totaly changed....

File: jans-cedarling/cedarling/src/init/service_config.rs (hunk 3 above)
Comment (olehbozhok) [reply jans-cedarling/cedarling/src/init/service_config.rs:64]: I feel OK to stop iterate =)
but thanks, I didn't know about that feature

File: jans-cedarling/cedarling/src/init/policy_store.rs (hunk 1 above)
Comment (abaghinyan) [reply jans-cedarling/cedarling/src/init/policy_store.rs]: `Err(err)=>` is just more clear then `else` to identify the error part

**Comment by olehbozhok** [comment]
Looks like implemented in the https://github.com/JanssenProject/jans/pull/10036

File: charts/janssen-all-in-one/README.md [hunk 6]
```diff
@@ -134,8 +134,8 @@ Kubernetes: `>=v1.22.0-0`
 | cnCouchbasePasswordFile | string | `"/etc/jans/conf/couchbase_password"` | Path to Couchbase password file |
 | cnCouchbaseSuperuserPasswordFile | string | `"/etc/jans/conf/couchbase_superuser_password"` | Path to Couchbase superuser password file |
 | cnDocumentStoreType | string | `"DB"` | Document store type to use for shibboleth files DB. |
-| cnGoogleApplicationCredentials | string | `"/etc/jans/conf/google-credentials.json"` | Base64 encoded service account. The sa must have roles/secretmanager.admin to use Google secrets and roles/spanner.databaseUser to use Spanner. Leave as this is a sensible default. |
-| cnPersistenceType | string | `"sql"` | Persistence backend to run Janssen with couchbase|hybrid|sql|spanner. |
+| cnGoogleApplicationCredentials | string | `"/etc/jans/conf/google-credentials.json"` | Base64 encoded service account. The sa must have roles/secretmanager.admin to use Google secrets. Leave as this is a sensible default. |
```
Comment (olehbozhok) [inline charts/janssen-all-in-one/README.md]: It is you change in `charts/janssen-all-in-one/README.md` ? 
I guess not, probably exit the way to remove all changes in MR for this file..

File: docs/cedarling/cedarling-policy-store.md [hunk 7]
```diff
@@ -13,36 +13,39 @@ tags:
...
-  "policy_stores": {
-      "some_unique_string_id": {
-          "name": "TestPolicy",
-          "description": "Once upon a time there was a Policy, that lived in a Store.",
-          "policies": { ... },
-          "schema": { ... },
-          "identity_source": { ... }
-      }
-  }
+  "name": "Jans::Store",
```
Comment (olehbozhok) [inline docs/cedarling/cedarling-policy-store.md:30]: Name only `Jans` and it is defined the `agama-lab`
and looks like you need to revert this change...

File: docs/cedarling/cedarling-policy-store.md [hunk 8]
```diff
@@ -155,94 +161,156 @@ This record contains the information needed to validate tokens from this issuer:
...
-        "840da5d85403f35ea76519ed1a18a33989f855bf1cf8": {
-            "description": "simple policy example",
-            "creation_date": "2024-09-20T17:22:39.996050",
-            "policy_content": "cedar_policy_encoded_in_base64"
-        }
+  "cedar_version": "v4.0.0",
+  "name": "Jans::store",
+  "description": "a simple policy example",
+  "policies": { 
+    "fbd921a51b...": {
```
Comment (olehbozhok) [inline docs/cedarling/cedarling-policy-store.md:252]: Please, left only one policy. Because `cedarling` support only one policy in store per file. And in `agama-lab` will do the same

**Comment by rmarinn** [comment]
Closing this PR in favor of a newer one that addresses the same issue with an improved approach. Please see: #10098
